doc/manpages/ccnpoke.1
'\" t
.\"     Title: ccnpoke
.\"    Author: [see the "AUTHOR" section]
.\" Generator: DocBook XSL Stylesheets v1.75.2 <http://docbook.sf.net/>
.\"      Date: 01/10/2014
.\"    Manual: \ \&
.\"    Source: \ \& 0.8.2
.\"  Language: English
.\"
.TH "CCNPOKE" "1" "01/10/2014" "\ \& 0\&.8\&.2" "\ \&"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
ccnpoke \- Read data from stdin and send it as a single ContentObject in response to an interest\&.
.SH "SYNOPSIS"
.sp
\fBccnpoke\fR [\-hflv] [\-k \fIkeyuri\fR] [\-p \fIn\fR] [\-t \fItype\fR] [\-V \fIseg\fR] [\-w \fItimeout\fR] [\-x \fIfreshness\fR] \fIccnxuri\fR
.SH "DESCRIPTION"
.sp
The \fBccnpoke\fR utility reads data from stdin and generates a single ContentOjbect, with attributes modfied by the given options, for the supplied CCNx URI, \fIccnxuri\fR, and sends that object in response to a matching interest\&.
.SH "OPTIONS"
.PP
\fB\-h\fR
.RS 4
Generate the help message\&.
.RE
.PP
\fB\-e\fR \fIfile\fR
.RS 4
Populate the optional ExtOpt element with the contents of the given file\&. The provided data should generally be ccnb encoded and limited in size, although ccnpeek does not enforce this\&.
.RE
.PP
\fB\-f\fR
.RS 4
Force sending the ContentObject to ccnd even if there is no matching interest\&. This should be used only for testing purposes\&.
.RE
.PP
\fB\-l\fR
.RS 4
Set the FinalBlockId based on the last component of the name\&.
.RE
.PP
\fB\-v\fR
.RS 4
Report on stdout the name of the ContentObject written\&.
.RE
.PP
\fB\-k\fR \fIkeyuri\fR
.RS 4
Specifies the name to be used in the KeyLocator of the ContentObject\&.
.RE
.PP
\fB\-p\fR \fIn\fR
.RS 4
Specifies the number of components of the URI that should be registered in the interest filter\&.
\fIn\fR=0 implies registering ccnx:/\&.
\fIn\fR
must not be greater than the number of components in the name\&. The default is to register the name as given, which will not include any derived version component implied by the
\fB\-V\fR
option\&.
.RE
.PP
\fB\-t\fR \fItype\fR
.RS 4
Specifies the type to be encoded in the ContentObject\&. Values are DATA (default), ENCR, GONE, KEY, LINK, NACK, as documented for ContentObjects\&.
.RE
.PP
\fB\-V\fR \fIseg\fR
.RS 4
Generate a version stamp, and use the specfied segment number as a suffix\&. If the segment number is "%00" then the FinalBlockId will also be set\&.
.RE
.PP
\fB\-w\fR \fItimeout\fR
.RS 4
Specifies the time to wait, in seconds, for an Interest to be received\&.
.RE
.PP
\fB\-x\fR \fIfreshness\fR
.RS 4
Specifies the Freshness, in seconds, to be encoded in the ContentObject\&.
.RE
.SH "EXIT STATUS"
.PP
\fB0\fR
.RS 4
Success
.RE
.PP
\fB1\fR
.RS 4
Failure (syntax or usage error, error sending ContentObject)
.RE
.SH "AUTHOR"
.sp
Michael Plass, Nick Briggs


NEWS
Changes between 0.8.1 and 0.8.2:
--------------------------------
This highlights some important changes since the most recent numbered release.
Refer to http://redmine.ccnx.org/versions/21 for more details on CCNx 0.8.2.

- This release features a refactored strategy layer, making it straightforward to add new
  strategies in a modular way.  Alongside the traditional "default" strategy there are
  two new examples, a "parallel" strategy that sends interests on all available faces
  and a "loadsharing" strategy that sends each arriving interest to just one of the
  available faces.  Strategies may be selected by prefix.  This selection may be
  done using new subcommands to the ccndc utility.

	* Feature #100929: Implement Junit test suites to facilitate bulk testing.
	* Feature #100983: strategy layer refactoring
	* Feature #100984: strategy labels on prefixes
	* Feature #100990: Restore SYNC_NOISY to INFO
	* Feature #100992: Add test-security to standard java tests
	* Feature #100993: Add test-versioning back into standard java tests
	* Feature #100994: Remove unused cruft from build.xml
	* Feature #101036: java sync should use Deque operations instead of Stack as recommended in the javadocs
	* Feature #101039: UDP mode for java netmanagers is not tested in standard testing
	* Feature #101053: ccncatchunks2 could use improved hole-filling and interest lifetime management
	* Bug #101030: build.xml setup of ccnr environment is incorrect
	* Bug #101033: Handling of extremely large values of MaxSuffixComponents is incorrect
	* Bug #101037: ccn_seqwriter library code should not limit the block size to 4096 bytes
	* Bug #101042: CcnFileProxy uses deprecated methods
	* Bug #101046: VLC 2.1.0 release necessitates change in plugin Makefile
	* Bug #101048: Do not rely on accept(2) returned addr for recognizing a unix-domain socket
	* Bug #101052: Unbreak objtree build target
	* Bug #101057: ccn_sign_content / ccn_chk_signing_params do not properly incorporate a timestamp in a template via ccn_signing_params
	* Bug #101058: ccndc in daemon mode does not exit cleanly when ccnd goes away
	* Bug #101059: testTimeouts a little too strict
	* Bug #101060: ccnd cache lookups should take a controllable amount of work
	* Bug #101061: ccnd minimum interest lifetime problem
	* Bug #101062: some ccnd logging missing check for logging enabled before calling expensive log procedure
	* Support #101047: Prepare for ccnx-0.8.2 release

Changes between 0.8.0 and 0.8.1:
--------------------------------
This highlights some important changes since the most recent numbered release.
Refer to http://redmine.ccnx.org/versions/20 for more details on CCNx 0.8.1.

- This release fixes a number of bugs, including a serious memory leak in ccnd.

	* Bug #101029: unbounded test log file growth
	* Bug #101032: ccnr should start at specified segment on checked start write
	* Bug #101034: No obvious documentation for how to start ccnr exists on website
	* Bug #101035: small leak in ccn_fetch should a realloc() fail
	* Bug #101038: Inconsistent encoding of zero-length name components
	* Bug #101043: ccnr r_store_lookup_backwards can fail
	* Bug #101044: valgrind reports memory leak in 0.8.0 ccnd
	* Bug #101049: ccnd can remove newly arrived content too quickly
	* Support #101045 Prepare for ccnx-0.8.1 release

Changes between 0.7.2 and 0.8.0:
--------------------------------
This highlights some important changes since the most recent numbered release.
Refer to http://redmine.ccnx.org/versions/17 for more details on CCNx 0.8.0.

- The use of symmetric key HMACs in place of public/private key signatures
  is now an option in both libraries.  In situations where it can be used,
  it provides a much higher performance alternative.
- The content store in ccnd has been improved.  It now uses a flatname
  representation for indexing, in the same way that ccnr does.  Also, the
  underlying skiplist implementation has moved into the library, so that
  applications may take advantage of it in the future.  In the process,
  some old ccnd bugs relating to the handling of extreme values of
  FreshnessSeconds were fixed.
- The more readable escaping convention in URIs is now the default.
  If for whatever reason you prefer to use only percent-escaping in URIs, use

	CFLAGS='-g -DCCN_URI_DEFAULT_ESCAPE=4' ./configure

  before typing make.  Note that both forms of escaping are accepted as input.
- An example rc startup script is now included, but needs to be installed by
  hand.  It is coded for Ubuntu Linux, but should be easily adaptable to other
  systems.
- This release drops support for Java JDK 1.5.

  * Issues Resolved:
	* Bug #100118: ccnd does not handle FreshnessSeconds=0 correctly.
	* Bug #100119: ccnd should handle large values of FreshnessSeconds
	* Bug #100558: test-security does not pass
	* Bug #100897: timeout values  should be consistently long throughout java code 
	* Bug #100966: Incorrect processing of CCNS_SYNC_SCOPE=3
	* Bug #100977: Problem with verify in CCNInputStream when trying to retrieve key
	* Bug #100987: Non-portable use of local keyword in Bourne shell scripts
	* Bug #100998: VLC 2.1.x block_New definition removed
	* Bug #100999: java BLOCK_AFTER_FIRST_SEGMENT doesn't work in all cases
	* Bug #101000: sync_diff code should reset differencing data when aborting or completing a comparison
	* Bug #101004: QueuedContentHandler has a sync error
	* Bug #101005: CCNx Android Services build out of sync with API changes to CCNx C-lib
	* Bug #101007: ccn_resolve_version() should distinguish between no response and finding an unversioned object.
	* Bug #101010: ccnseqwriter command busy-loops when its ccnd goes away
	* Bug #101011: Compile warning when building new symmetric key code with openssl 0.9.8
	* Bug #101012: Improve recovery from node fetch timeout in java sync
	* Bug #101013: node fetch timeout in java sync implementation should be longer and settable

csrc/cmd/ccnpoke.c
/**
 * @file ccnpoke.c
 * Injects one chunk of data from stdin into ccn.
 *
 * A CCNx command-line utility.
 *
 * Copyright (C) 2008-2013 Palo Alto Research Center, Inc.
 *
 * This work is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License version 2 as published by the
 * Free Software Foundation.
 * This work is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
 * for more details. You should have received a copy of the GNU General Public
 * License along with this program; if not, write to the
 * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
 * Boston, MA 02110-1301, USA.
 */
#include <errno.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <ccn/ccn.h>
#include <ccn/uri.h>
#include <ccn/keystore.h>
#include <ccn/signing.h>

static ssize_t
read_full(int fd, unsigned char *buf, size_t size)
{
    size_t i;
    ssize_t res = 0;
    for (i = 0; i < size; i += res) {
        res = read(fd, buf + i, size - i);
        if (res == -1) {
            if (errno == EAGAIN || errno == EINTR)
                res = 0;
            else
                return(res);
        }
        else if (res == 0)
            break;
    }
    return(i);
}

static void
usage(const char *progname)
{
    fprintf(stderr,
            "%s [-hflv] [-k key_uri] [-t type] [-V seg] [-w timeout] [-x freshness_seconds]"
            " ccnx:/some/place\n"
            " Reads data from stdin and sends it to the local ccnd"
            " as a single ContentObject under the given URI\n"
            "  -h - print this message and exit\n"
            "  -e file - extopt from supplied file\n"
            "  -f - force - send content even if no interest received\n"
            "  -l - set FinalBlockId from last segment of URI\n"
            "  -v - verbose\n"
            "  -k key_uri - use this name for key locator\n"
            "  -p n - limit registration to n (>=0) components of the given URI in the interest filter.\n"
            "  -t ( DATA | ENCR | GONE | KEY | LINK | NACK ) - set type\n"
            "  -V seg - generate version, use seg as name suffix\n"
            "  -w seconds - fail after this long if no interest arrives\n"
            "  -x seconds - set FreshnessSeconds\n"
            , progname);
    exit(1);
}

enum ccn_upcall_res
incoming_interest(
    struct ccn_closure *selfp,
    enum ccn_upcall_kind kind,
    struct ccn_upcall_info *info)
{
    struct ccn_charbuf *cob = selfp->data;
    int res;
    
    switch (kind) {
        case CCN_UPCALL_FINAL:
            break;
        case CCN_UPCALL_INTEREST:
            if (ccn_content_matches_interest(cob->buf, cob->length,
                    1, NULL,
                    info->interest_ccnb, info->pi->offset[CCN_PI_E],
                    info->pi)) {
                res = ccn_put(info->h, cob->buf, cob->length);
                if (res >= 0) {
                    selfp->intdata = 1;
                    ccn_set_run_timeout(info->h, 0);
                    return(CCN_UPCALL_RESULT_INTEREST_CONSUMED);
                }
            }
            break;
        default:
            break;
    }
    return(CCN_UPCALL_RESULT_OK);
}

int
main(int argc, char **argv)
{
    const char *progname = argv[0];
    struct ccn *ccn = NULL;
    struct ccn_charbuf *name = NULL;
    struct ccn_charbuf *pname = NULL;
    struct ccn_charbuf *temp = NULL;
    struct ccn_charbuf *extopt = NULL;
    long expire = -1;
    int versioned = 0;
    size_t blocksize = 8*1024;
    int status = 0;
    int res;
    ssize_t read_res;
    unsigned char *buf = NULL;
    enum ccn_content_type content_type = CCN_CONTENT_DATA;
    struct ccn_closure in_interest = {.p=&incoming_interest};
    const char *postver = NULL;
    const char *key_uri = NULL;
    int force = 0;
    int verbose = 0;
    int timeout = -1;
    int setfinal = 0;
    int prefixcomps = -1;
    int fd;
    struct ccn_signing_params sp = CCN_SIGNING_PARAMS_INIT;
    
    while ((res = getopt(argc, argv, "e:fhk:lvV:p:t:w:x:")) != -1) {
        switch (res) {
            case 'e':
                if (extopt == NULL)
                    extopt = ccn_charbuf_create();
                fd = open(optarg, O_RDONLY);
                if (fd < 0) {
                    perror(optarg);
                    exit(1);
                }
                for (;;) {
                    read_res = read(fd, ccn_charbuf_reserve(extopt, 64), 64);
                    if (read_res <= 0)
                        break;
                    extopt->length += read_res;
                }
                if (read_res < 0)
                    perror(optarg);
                close(fd);
                break;
            case 'f':
                force = 1;
                break;
            case 'l':
                setfinal = 1; // set FinalBlockID to last comp of name
                break;
            case 'k':
                key_uri = optarg;
                break;
            case 'p':
                prefixcomps = atoi(optarg);
                if (prefixcomps < 0)
                    usage(progname);
                break;
            case 'x':
                expire = atol(optarg);
                if (expire <= 0)
                    usage(progname);
                break;
            case 'v':
                verbose = 1;
                break;
            case 'V':
                versioned = 1;
                postver = optarg;
                if (0 == memcmp(postver, "%00", 3))
                    setfinal = 1;
                break;
            case 'w':
                timeout = atol(optarg);
                if (timeout <= 0)
                    usage(progname);
                timeout *= 1000;
                break;
            case 't':
                if (0 == strcasecmp(optarg, "DATA")) {
                    content_type = CCN_CONTENT_DATA;
                    break;
                }
                if (0 == strcasecmp(optarg, "ENCR")) {
                    content_type = CCN_CONTENT_ENCR;
                    break;
                }
                if (0 == strcasecmp(optarg, "GONE")) {
                    content_type = CCN_CONTENT_GONE;
                    break;
                }
                if (0 == strcasecmp(optarg, "KEY")) {
                    content_type = CCN_CONTENT_KEY;
                    break;
                }
                if (0 == strcasecmp(optarg, "LINK")) {
                    content_type = CCN_CONTENT_LINK;
                    break;
                }
                if (0 == strcasecmp(optarg, "NACK")) {
                    content_type = CCN_CONTENT_NACK;
                    break;
                }
                content_type = atoi(optarg);
                if (content_type > 0 && content_type <= 0xffffff)
                    break;
                fprintf(stderr, "Unknown content type %s\n", optarg);
                /* FALLTHRU */
            default:
            case 'h':
                usage(progname);
                break;
        }
    }
    argc -= optind;
    argv += optind;
    if (argv[0] == NULL)
        usage(progname);
    name = ccn_charbuf_create();
    res = ccn_name_from_uri(name, argv[0]);
    if (res < 0) {
        fprintf(stderr, "%s: bad ccn URI: %s\n", progname, argv[0]);
        exit(1);
    }
    if (argv[1] != NULL)
        fprintf(stderr, "%s warning: extra arguments ignored\n", progname);
    
    /* Preserve the original prefix, in case we add versioning,
     * but trim it down if requested for the interest filter registration
     */
    pname = ccn_charbuf_create();
    ccn_charbuf_append(pname, name->buf, name->length);
    if (prefixcomps >= 0) {
        res = ccn_name_chop(pname, NULL, prefixcomps);
        if (res < 0) {
            fprintf(stderr, "%s: unable to trim name to %d component%s.\n",
                    progname, prefixcomps, prefixcomps == 1 ? "" : "s");
            exit(1);
        }
    }
    /* Connect to ccnd */
    ccn = ccn_create();
    if (ccn_connect(ccn, NULL) == -1) {
        perror("Could not connect to ccnd");
        exit(1);
    }

    /* Read the actual user data from standard input */
    buf = calloc(1, blocksize);
    read_res = read_full(0, buf, blocksize);
    if (read_res < 0) {
        perror("read");
        read_res = 0;
        status = 1;
    }
        
    /* Tack on the version component if requested */
    if (versioned) {
        res = ccn_create_version(ccn, name, CCN_V_REPLACE | CCN_V_NOW | CCN_V_HIGH, 0, 0);
        if (res < 0) {
            fprintf(stderr, "%s: ccn_create_version() failed\n", progname);
            exit(1);
        }
        if (postver != NULL) {
            res = ccn_name_from_uri(name, postver);
            if (res < 0) {
                fprintf(stderr, "-V %s: invalid name suffix\n", postver);
                exit(0);
            }
        }
    }
    temp = ccn_charbuf_create();
    
    /* Ask for a FinalBlockID if appropriate. */
    if (setfinal)
        sp.sp_flags |= CCN_SP_FINAL_BLOCK;
    
    if (res < 0) {
        fprintf(stderr, "Failed to create signed_info (res == %d)\n", res);
        exit(1);
    }
    
    /* Set content type */
    sp.type = content_type;
    
    /* Set freshness */
    if (expire >= 0) {
        if (sp.template_ccnb == NULL) {
            sp.template_ccnb = ccn_charbuf_create();
            ccnb_element_begin(sp.template_ccnb, CCN_DTAG_SignedInfo);
        }
        else if (sp.template_ccnb->length > 0) {
            sp.template_ccnb->length--;
        }
        ccnb_tagged_putf(sp.template_ccnb, CCN_DTAG_FreshnessSeconds, "%ld", expire);
        sp.sp_flags |= CCN_SP_TEMPL_FRESHNESS;
        ccnb_element_end(sp.template_ccnb);
    }
    
    /* Set key locator, if supplied */
    if (key_uri != NULL) {
        struct ccn_charbuf *c = ccn_charbuf_create();
        res = ccn_name_from_uri(c, key_uri);
        if (res < 0) {
            fprintf(stderr, "%s is not a valid ccnx URI\n", key_uri);
            exit(1);
        }
        if (sp.template_ccnb == NULL) {
            sp.template_ccnb = ccn_charbuf_create();
            ccnb_element_begin(sp.template_ccnb, CCN_DTAG_SignedInfo);
        }
        else if (sp.template_ccnb->length > 0) {
            sp.template_ccnb->length--;
        }
        ccnb_element_begin(sp.template_ccnb, CCN_DTAG_KeyLocator);
        ccnb_element_begin(sp.template_ccnb, CCN_DTAG_KeyName);
        ccn_charbuf_append(sp.template_ccnb, c->buf, c->length);
        ccnb_element_end(sp.template_ccnb);
        ccnb_element_end(sp.template_ccnb);
        sp.sp_flags |= CCN_SP_TEMPL_KEY_LOCATOR;
        ccnb_element_end(sp.template_ccnb);
        ccn_charbuf_destroy(&c);
    }

    if (extopt != NULL && extopt->length > 0) {
        if (sp.template_ccnb == NULL) {
            sp.template_ccnb = ccn_charbuf_create();
            ccnb_element_begin(sp.template_ccnb, CCN_DTAG_SignedInfo);
        }
        else if (sp.template_ccnb->length > 0) {
            sp.template_ccnb->length--;
        }
        ccnb_append_tagged_blob(sp.template_ccnb, CCN_DTAG_ExtOpt,
                                extopt->buf, extopt->length);
        sp.sp_flags |= CCN_SP_TEMPL_EXT_OPT;
        ccnb_element_end(sp.template_ccnb);
    }
    
    /* Create the signed content object, ready to go */
    temp->length = 0;
    res = ccn_sign_content(ccn, temp, name, &sp, buf, read_res);
    if (res != 0) {
        fprintf(stderr, "Failed to encode ContentObject (res == %d)\n", res);
        exit(1);
    }
    if (read_res == blocksize) {
        read_res = read_full(0, buf, 1);
        if (read_res == 1) {
            fprintf(stderr, "%s: warning - truncated data\n", argv[0]);
            status = 1;
        }
    }
    free(buf);
    buf = NULL;
    if (force) {
        /* At user request, send without waiting to see an interest */
        res = ccn_put(ccn, temp->buf, temp->length);
        if (res < 0) {
            fprintf(stderr, "ccn_put failed (res == %d)\n", res);
            exit(1);
        }
    }
    else {
        in_interest.data = temp;
        /* Set up a handler for interests */
        res = ccn_set_interest_filter(ccn, pname, &in_interest);
        if (res < 0) {
            fprintf(stderr, "Failed to register interest (res == %d)\n", res);
            exit(1);
        }
        res = ccn_run(ccn, timeout);
        if (in_interest.intdata == 0) {
            if (verbose)
                fprintf(stderr, "Nobody's interested\n");
            exit(1);
        }
    }
    
    if (verbose) {
        struct ccn_charbuf *uri = ccn_charbuf_create();
        uri->length = 0;
        ccn_uri_append(uri, name->buf, name->length, 1);
        printf("wrote %s\n", ccn_charbuf_as_string(uri));
        ccn_charbuf_destroy(&uri);
    }
    ccn_destroy(&ccn);
    ccn_charbuf_destroy(&name);
    ccn_charbuf_destroy(&pname);
    ccn_charbuf_destroy(&temp);
    ccn_charbuf_destroy(&sp.template_ccnb);
    ccn_charbuf_destroy(&extopt);
    exit(status);
}
